src/webui/api/searchcontroller.cpp

```diff
@@ -0,0 +1,344 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "searchcontroller.h"
+
+#include <QRegularExpression>
+#include <QSharedPointer>
+#include <QTimer>
+
+#include "base/logger.h"
+#include "base/search/searchhandler.h"
+#include "base/utils/foreignapps.h"
+#include "base/utils/random.h"
+#include "base/utils/string.h"
+#include "apierror.h"
+#include "isessionmanager.h"
+
+class SearchPluginManager;
+
+using SearchHandlerPtr = QSharedPointer<SearchHandler>;
+using SearchHandlerDict = QHash<QString, SearchHandlerPtr>;
+
+void SearchController::startAction()
+{
+    checkParams({"pattern", "category", "plugins"});
+
+    if (Utils::ForeignApps::pythonInfo().version.majorNumber() <= 0)
+        throw APIError(APIErrorType::Conflict, "Python must be installed to use the Search Engine.");
+
+    const QString pattern = params()["pattern"].trimmed();
+    const QString category = params()["category"].trimmed();
+    const QString plugins = params()["plugins"].trimmed();
+    const QString pluginsLower = plugins.toLower();
+
+    QStringList pluginsToUse;
+    if (pluginsLower == "all")
+        pluginsToUse = SearchPluginManager::instance()->allPlugins();
+    else if ((pluginsLower == "enabled") || (pluginsLower == "multi"))
+        pluginsToUse = SearchPluginManager::instance()->enabledPlugins();
+    else
+        pluginsToUse << plugins;
+
+    const QString id = generateSearchId();
+
+    auto searchHandlers = sessionManager()->session()->getData(QLatin1String("searchHandlers")).value<SearchHandlerDict>();
+    SearchHandlerPtr searchHandler {SearchPluginManager::instance()->startSearch(pattern, category, pluginsToUse)};
+    searchHandlers.insert(id, searchHandler);
+    sessionManager()->session()->setData(QLatin1String("searchHandlers"), QVariant::fromValue(searchHandlers));
+
+    const int hour = 3600000;
+    QTimer::singleShot(hour, this, [id, this] () {searchTimeout(id);});
```

Comment:
I guess `SearchController::startAction()` could be invoked as many times as user want, the problem is QTimer could be a limited resource on some OS, from qtimer doc:
>Some operating systems limit the number of timers that may be used; Qt tries to work around these limitations.

So maybe it is better to have a QMap that records the (id, start time) pair and have a single qtimer that fires every hour.

Reply:
@Chocobo1, please see https://github.com/qbittorrent/qBittorrent/pull/8584#discussion_r201962657

Reply:
Now that I'm thinking about implementation, I'm inclined to set a limit of 10 concurrent searches.  It would protect against hitting the QTimer limit and was already being suggested.